src/modules/mavlink/mavlink_stream.cpp

```diff
@@ -128,3 +128,200 @@ MavlinkStream::update(const hrt_abstime &t)
+	}
+}
+
+/**
+ * Register stream object to poll list of orbs
+ */
+int
+MavlinkStreamPoll::register_orbs(uint16_t stream_id, ORB_ID *orbs, int cnt)
+{
+	if (!orbs || cnt <= 0) {
+		return PX4_OK;
+	}
+
+	for (int i = 0; i < cnt; i++) {
+
+		// Increase reqs capacity if necessary
+		while (_reqs_count >= _reqs_capacity) {
+			_reqs_capacity = _reqs_capacity + 16;
+			_reqs = (MavStreamOrbPollReq *)realloc(_reqs, _reqs_capacity * sizeof(MavStreamOrbPollReq));
+
+			if (!_reqs) {
+				PX4_ERR("failed to allocate memory for orb poll reqs");
+				return PX4_ERROR;
+			}
+		}
+
+
+		MavStreamOrbPollReq *req = &_reqs[_reqs_count];
+		req->stream_id = stream_id;
+		req->orb_id = orbs[i];
+		req->interval = -1;
+		_reqs_count++;
+
+		// Check if the orb is already in the list
+		bool exists = false;
+
+		for (int j = 0; j < _count; j++) {
+			if (_orbs[j].orb_id == orbs[i]) {
+				_orbs[j].usage_count++;
+				exists = true;
+				break;
+			}
+		}
+
+		// if new orb is requested add it to the list and subscribe
+		if (!exists) {
+			while (_count >= _capacity) {
+				_capacity = _capacity + 16;
+				_orbs = (MavStreamPollItem *)realloc(_orbs, _capacity * sizeof(MavStreamPollItem));
+				_fds = (orb_poll_struct_t *)realloc(_fds, _capacity * sizeof(orb_poll_struct_t));
+
+				if (!_fds || !_orbs) {
+					PX4_ERR("failed to allocate memory for orb poll items");
+					return PX4_ERROR;
+				}
+			}
+
+			_orbs[_count].orb_id = orbs[i];
+			_orbs[_count].usage_count = 1;
+			_orbs[_count].fd = orb_subscribe(get_orb_meta(orbs[i]));
+			_count++;
+		}
+	}
+
+	// Update fds
+	for (int i = 0; i < _count; i++) {
+		_fds[i].fd = _orbs[i].fd;
+		_fds[i].events = POLLIN;
+		_fds[i].revents = 0;
+	}
+
+	return PX4_OK;
+}
+
+/**
+ * Unregister stream object from orbs polling
+ */
+int
+MavlinkStreamPoll::unregister_orbs(uint16_t stream_id)
+{
+	// Remove all requests from the unregistering stream
+	for (int i = 0; i < _reqs_count; i++) {
+		if (_reqs[i].stream_id == stream_id) {
+			for (int j = 0; j < _count; j++) {
+				if (_orbs[j].orb_id == _reqs[i].orb_id) {
+					_orbs[j].usage_count--;
+
+					// If this was the last request for the orb,
+					// unsubscribe and remove from orbs list
+					if (_orbs[j].usage_count == 0) {
+						orb_unsubscribe(_orbs[j].fd);
+						_orbs[j] = _orbs[--_count];
+					}
+				}
+			}
+
+			// Remove request from reqs list
+			_reqs[i] = _reqs[--_reqs_count];
+		}
+	}
+
+	// Update fds
+	for (int i = 0; i < _count; i++) {
+		_fds[i].fd = _orbs[i].fd;
+		_fds[i].events = POLLIN;
+		_fds[i].revents = 0;
+	}
+
+	return PX4_OK;
+}
+
+/**
+ * Set stream update interval to adjust orb polling accordingly
+ */
+int
+MavlinkStreamPoll::set_interval(uint16_t stream_id, int interval_ms)
+{
+	// Update all requests from the stream
+	for (int i = 0; i < _reqs_count; i++) {
+		if (_reqs[i].stream_id == stream_id) {
+			_reqs[i].interval = interval_ms;
+		}
+	}
+
+	// Update interval for all orb subscriptions and find the maximum interval
```

Comment:
Shouldn't we look for mininum interval? If someone streams the uorb with 10ms and another one with 20ms, whe should wake up every 10ms I suppose. This gets hairy if the times are 10 and 15, then the interval should be set to 5 (the largest common denominator...)

Reply:
Fixed to poll same orb with different intervals to maintain proper rates